### ISSUE TITLE
Define default value to missing Provider

### ISSUE THREAD

**Issue by daniel-v** [dart-lang/angular#343]
## Situation:

I have a simple component, that depends on a injected String value:

```dart
const A_TOKEN = const OpaqueToken('some.package.lib.token');

@Component(
    selector: 'my-cmp',
    template: '{{injectedString}}'
)
class MyComponent {
  final String injectedString;
  MyComponent(@Inject(A_TOKEN) @SkipSelf() this.injectedString);
}
```

Just by using this component, the DI will fail, as there there is no provider defined anywhere in the application and `MyComponent` expects a parent injector to be able to provide a value to it.

## Problem:
Should such a "failure" occur, I still would like to fall back to a default value.

### Workarounds:

I can achieve the desired result in multiple different ways. Eg:

```dart
class MyComponent {
  final String injectedString;
  final Injector _injector;
  MyComponent(@SkipSelf() this._injector) // parent injector
      : injectedString = _injector.get(A_TOKEN, 'defaultValue'); // supported functionality!
}
```

or

```dart
class MyComponent {
  final String injectedString;
  MyComponent(@Inject(A_TOKEN) @SkipSelf() @Optional() String str)
      : injectedString = str ?? 'defaultValue';
}
```

or can be solved with a factory.

## What I'm looking for:

A way to define a fallback value through Provider definitions. An imaginary way I'd like to do it is:

```dart
@Component(
    selector: 'my-cmp',
    template: '{{injectedString}}',
    viewProviders: const [
      const DefaultProvider(A_TOKEN, useValue: 'defaultValue')
    ]
)
class MyComponent {
  @Inject(A_TOKEN) 
  @SkipSelf()                 // expect A_TOKEN to be defined by a parent
  String injectedString;  // but fall back to DefaultProvider should it not be
}
```

## How is this different:

1. DI would only fail if `A_TOKEN` has no Provider OR DefaultProvider
2. If a Provider is defined, DefaultProvider is ignored
3. Don't need to touch code to define this behavior

### Some reasoning:

I am writing re-usable, fairly customizable components and _they come with some "sensible defaults"_. These components are **context-aware**, their place of use influence their behavior or look.

Angular2's concept of providing context to components happen through either Providers or the Component Tree. Both are facilitated by the DI. 

Therefore it makes sense to me that supporting fallback/default provider values should be supported by the framework itself.

**Comment by matanlurey** [comment]
We've had similar requests internally, and as you've already imagined it can be tough to get the semantics right. You also don't want to have performance regressions, for example, having 500 components on a page all checking for a provider before using a default, etc.

That being said, I don't think this is a bad idea. What about a different syntax though?

```dart
class MyComponent {
  @Inject(A_TOKEN)
  @Optional.defaultTo('someValue')
  final String injectedString;

  // Or

  @Inject(A_TOKEN)
  @Optional.defaultToFactory(getSomeValue)
  MyComponent(this.injectedString);
}
```

That would make the implementation quite a bit simpler, as it would mainly just be sugar for what you're already doing today.

**Comment by daniel-v** [comment]
When I proposed the use of `DefaultProvider`, I also had in mind the searchability of such providers in case I needed to find where defaults are being used in my project. Using aforementioned syntax also satisfy that, though a bit more tedious if there are multiple versions of it: `.defaultTo` and `.defaultToFactory`.

The proposed syntactic sugar does satisfy the example - `injectedString` gets a value.

I am looking at this one from the workarounds:

```dart
  MyComponent(@SkipSelf() this._injector) // parent injector
      : injectedString = _injector.get(A_TOKEN, 'defaultValue'); 
```

**Question 1:**

What would happen, if it's not that simple to provide? Can I declare dependencies from the DI tree to the `defaultToFactory`? Something like:

```dart
abstract class StringProducer {
  String produce();
}

String getSomeValue(StringProducer producer) =>
  producer.produce();

class MyComponent {  
  @Inject(A_TOKEN)
  @Optional.defaultToFactory(getSomeValue, deps: const [ 
    const Provider( StringProducer, useClass: RandomStringProducer ) 
  ])
  final string injectedString;
  MyComponent(this.injectedString);
}

```

Right now, this is what I'd write:

```dart
MyComponent(
    @Inject() @SkipSelf() Injector parentInjector,
    @Inject() Injector componentInjector)
    : injectedString = _injector.get(
        A_TOKEN,
        componentInjector.get(StringProducer).produce() );
```

*Potential issue I see:*
the `_injector` parameter is marked with `@SkipSelf()`, need to make sure that the component injector is explicitly provided again.

**Simplification**

Most uses I see to this feature does work with simple values. A String, or a bool or an instance of a class. I presume it's more complicated to implement that "syntactic sugar" this way.  Maybe the following is simple enough and provides the power we might need in some cases:

```dart
@Inject(A_TOKEN)
@Optional.defaultToFactory(getSomeValue)
MyComponent(this.injectedString);

String getSomeValue(Injector componentInjector) =>
  injector.get(StringProducer).produce();
```

**Potential issue**:
Which injector is this? 
In other words there is no support for `@Self()`, `@SkipSelf()` and `@Host()` for the `componentInjector`. Should the interface of `Injector` be extended with `Injector get parent`, SkipSelf is implemented.
An optional bool parameter to `Injector.get()` could prevent the Injector instance to start walking the tree. This would provide implementation for `Self()`. 
`injectorInstance.parent.get(TOKEN, doNotWalk: true)` could do for `@Host`.

I suspect, this would also provide fine-grained control over performance trade-offs while being powerful enough to support the "not-so-simple" use-cases.

**Comment by matanlurey** [comment]
This has been accepted (in spirit) :) Not sure about exact syntax or semantics yet.

**Comment by matanlurey** [comment]
I think we're unlikely to act on this immediately, unfortunately.

It makes code generation much much more complex.

**Comment by sergiubologa** [comment]
Hello. Now with Ivy, is there a better alternative to this problem?
